WalletWasabi.Fluent/ViewModels/AddWallet/ConnectHardwareWalletViewModel.cs

```diff
@@ -0,0 +1,205 @@
+				var path = _walletManager.WalletDirectories.GetWalletFilePaths(_walletName).walletFilePath;
+
+				_walletManager.AddWallet(KeyManager.CreateNewHardwareWalletWatchOnly(fingerPrint, extPubKey, path));
+
+				// Close dialog
+				ClearNavigation();
+			}
+			catch (Exception ex)
+			{
+				// TODO: Notify the user about the error
+				Logger.LogError(ex);
+
+				// Restart detection
+				StartDetection();
+			}
+		}
+
+		private async Task<ExtPubKey> GetXpubAsync(HardwareWalletViewModel wallet)
+		{
+			using var cts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
+			var tryCounter = 1;
+
+			while (true)
+			{
+				try
+				{
+					return await _hwiClient.GetXpubAsync(wallet.HardwareWalletInfo.Model, wallet.HardwareWalletInfo.Path, KeyManager.DefaultAccountKeyPath, cts.Token);
+				}
+				catch (OperationCanceledException)
+				{
+					throw;
+				}
+				catch (Exception ex)
+				{
+					if (tryCounter++ >= 3)
+					{
+						throw;
+					}
+				}
+			}
+		}
+
+		private Task StopDetection() => Task.Run(() =>
+		{
+			_searchHardwareWalletCts.Cancel();
+
+			while (_detectionTask is { } && !_detectionTask.IsCompleted)
+			{
+				Thread.Sleep(100);
+			}
+		});
+
+		private void StartDetection()
+		{
+			_searchHardwareWalletCts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
+			_detectionTask = new Task(HardwareWalletDetectionAsync);
+			_detectionTask.Start();
+		}
+
+		private async void HardwareWalletDetectionAsync()
+		{
+			while (!_searchHardwareWalletCts.IsCancellationRequested)
+			{
+				var sw = new Stopwatch();
+				sw.Start();
+
+				try
+				{
+					// Reset token
+					_searchHardwareWalletCts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
+
+					var detectedHardwareWallets = (await _hwiClient.EnumerateAsync(_searchHardwareWalletCts.Token)).Select(x => new HardwareWalletViewModel(x)).ToList();
+
+					// Remove wallets that are already added to software
+					var walletsToRemove = detectedHardwareWallets.Where(wallet => _walletManager.GetWallets().Any(x => x.KeyManager.MasterFingerprint == wallet.HardwareWalletInfo.Fingerprint));
+					detectedHardwareWallets.RemoveMany(walletsToRemove);
+
+					// Remove disconnected hardware wallets from the list
+					HardwareWallets.RemoveMany(HardwareWallets.Except(detectedHardwareWallets));
+
+					// Remove detected wallets that are already in the list.
+					detectedHardwareWallets.RemoveMany(HardwareWallets);
+
+					// All remained detected hardware wallet is new so add.
+					HardwareWallets.AddRange(detectedHardwareWallets);
+				}
+				catch (Exception ex)
+				{
+					if (!(ex is OperationCanceledException))
+					{
+						Logger.LogError(ex);
+					}
+				}
+
+				// Too fast enumeration causes the detected hardware wallets cannot provide the fingerprint.
+				// Wait at least 5 seconds between two enumerations.
+				sw.Stop();
+				if (sw.Elapsed.Milliseconds < 5000)
+				{
+					Thread.Sleep(5000 - sw.Elapsed.Milliseconds);
```

Comment:
I think it might be a bad idea to use Thread.Sleep instead of Task.Delay.

It means this Thread is completely unavailable for use in the TaskPool for 5 seconds... Task.Delay would allow the thread to be used for something else.

Reply:
Done